File: youtube_line.py
import cv2
import yt_dlp
import numpy as np

# YouTube URL에서 비디오 스트림 가져오기
def get_video_stream(youtube_url):
    ydl_opts = {'format': 'best[ext=mp4]'}  # mp4 형식의 최고 화질 스트림
    with yt_dlp.YoutubeDL(ydl_opts) as ydl:
        info_dict = ydl.extract_info(youtube_url, download=False)  # 스트림 URL 가져오기
        video_url = info_dict['url']
        fps = info_dict.get('fps', 30)  # 동영상 FPS 가져오기 (기본값 30)
    return video_url, fps

# 도로 직선 검출
def detect_road_lanes(frame):
    # 그레이스케일로 변환
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

    # 가우시안 블러 적용 (노이즈 제거)
    blurred = cv2.GaussianBlur(gray, (5, 5), 0)

    # Canny Edge Detection 적용
    edges = cv2.Canny(blurred, 50, 150)

    # ROI (관심 영역) 설정
    height, width = edges.shape
    mask = np.zeros_like(edges)

    # 도로에 해당하는 다각형 영역 정의
    polygon = np.array([[
        (width * 0.1, height),  # 왼쪽 하단
        (width * 0.9, height),  # 오른쪽 하단
        (width * 0.6, height * 0.6),  # 오른쪽 위
        (width * 0.4, height * 0.6)   # 왼쪽 위
    ]], dtype=np.int32)

    # 관심 영역에 대한 마스크 생성
    cv2.fillPoly(mask, polygon, 255)
    roi_edges = cv2.bitwise_and(edges, mask)

    # 허프 변환 직선 검출
    lines = cv2.HoughLinesP(
        roi_edges, 
        rho=1, 
        theta=np.pi / 180, 
        threshold=50, 
        minLineLength=100, 
        maxLineGap=50
    )

    # 도로 위 직선 그리기
    line_image = np.zeros_like(frame)
    if lines is not None:
        for line in lines:
            x1, y1, x2, y2 = line[0]
            cv2.line(line_image, (x1, y1), (x2, y2), (0, 255, 0), 5)  # 초록색 직선

    # 원본 프레임에 직선 덮어쓰기
    combined = cv2.addWeighted(frame, 0.8, line_image, 1, 1)
    return combined

# 비디오 처리
def process_video(video_url, fps):
    cap = cv2.VideoCapture(video_url)
    if not cap.isOpened():
        print("Error: Cannot open video stream.")
        return

    frame_delay = int(1000 / fps)  # 프레임 딜레이 계산 (ms 단위)

    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            print("Finished processing video.")
            break

        # 도로 직선 검출 및 시각화
        frame_with_lanes = detect_road_lanes(frame)

        # 결과 프레임을 표시
        cv2.imshow("Road Lane Detection", frame_with_lanes)

        # 'q'를 눌러 종료
        if cv2.waitKey(frame_delay) & 0xFF == ord('q'):  # FPS에 맞춘 딜레이 적용
            break

    cap.release()
    cv2.destroyAllWindows()

# 메인 함수
if __name__ == "__main__":
    youtube_url = "https://www.youtube.com/watch?v=icc0a3yAwnQ"  # 유튜브 동영상 URL
    video_url, fps = get_video_stream(youtube_url)
    fps = 30  # FPS를 강제로 30으로 설정
    print(f"Video FPS: {fps}")
    process_video(video_url, fps)

File: youtube_line.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <string>

using namespace cv;
using namespace std;

// YouTube 동영상의 실제 스트림 URL을 가져오는 함수
string getYouTubeStreamUrl(const string& youtubeUrl) {
    // yt-dlp 명령어를 실행하여 스트림 URL 가져오기
    string command = "yt-dlp -f 'best[ext=mp4]' --get-url \"" + youtubeUrl + "\" 2> yt-dlp-error.log > video_url.txt";
    cout << "Executing command: " << command << endl;
    int ret = system(command.c_str());  // yt-dlp 명령어 실행

    if (ret != 0) {
        cerr << "Error: yt-dlp 명령어 실행 실패. yt-dlp-error.log를 확인하세요." << endl;
        return "";
    }

    ifstream urlFile("video_url.txt");  // 스트림 URL이 저장된 파일 열기
    string videoUrl;
    if (urlFile.is_open()) {
        getline(urlFile, videoUrl);  // 첫 번째 줄에서 URL 읽기
        urlFile.close();
    } else {
        cerr << "Error: 파일에서 비디오 URL을 읽을 수 없습니다." << endl;
        return "";
    }
    return videoUrl;  // 스트림 URL 반환
}

// 도로 차선 검출 함수
void detectLaneLines(Mat& frame) {
    Mat gray, blur, edges, mask, roi;

    // 1. 이미지를 그레이스케일로 변환
    cvtColor(frame, gray, COLOR_BGR2GRAY);

    // 2. 가우시안 블러 적용 (노이즈 제거)
    GaussianBlur(gray, blur, Size(5, 5), 0);

    // 3. Canny 에지 검출
    Canny(blur, edges, 50, 150);

    // 4. 관심 영역(ROI) 설정
    mask = Mat::zeros(edges.size(), edges.type());
    Point points[1][4] = {
        {Point(frame.cols * 0.1, frame.rows), Point(frame.cols * 0.45, frame.rows * 0.6),
         Point(frame.cols * 0.55, frame.rows * 0.6), Point(frame.cols * 0.9, frame.rows)}};
    const Point* ppt[1] = {points[0]};
    int npt[] = {4};
    fillPoly(mask, ppt, npt, 1, Scalar(255, 255, 255));  // 다각형 영역 채우기

    // 에지 이미지와 ROI 마스크의 AND 연산
    bitwise_and(edges, mask, roi);

    // 5. 허프 변환을 사용한 선 검출
    vector<Vec4i> lines;
    HoughLinesP(roi, lines, 1, CV_PI / 180, 50, 50, 10);

    // 검출된 선을 원본 프레임에 그림
    for (size_t i = 0; i < lines.size(); i++) {
        Vec4i l = lines[i];
        line(frame, Point(l[0], l[1]), Point(l[2], l[3]), Scalar(0, 255, 0), 3, LINE_AA);
    }
}

// 비디오 스트림 처리 함수
void processVideo(const string& videoUrl) {
    VideoCapture cap(videoUrl);  // 비디오 스트림 열기
    if (!cap.isOpened()) {
        cerr << "Error: 비디오 스트림을 열 수 없습니다." << endl;
        return;
    }

    Mat frame;
    int fps = static_cast<int>(cap.get(CAP_PROP_FPS));  // FPS 가져오기
    if (fps == 0) fps = 30;  // 기본 FPS 설정

    int frameDelay = 1000 / fps;  // 프레임 딜레이(ms)

    while (true) {
        cap >> frame;  // 프레임 읽기
        if (frame.empty()) {
            cout << "비디오 처리가 완료되었습니다." << endl;
            break;
        }

        // 도로 차선 검출
        detectLaneLines(frame);

        // 결과를 화면에 표시
        imshow("YouTube Video Stream - Lane Detection", frame);

        // 'q' 키를 누르면 종료
        if (waitKey(frameDelay) == 'q') {
            break;
        }
    }

    cap.release();  // 비디오 스트림 해제
    destroyAllWindows();  // 모든 창 닫기
}

// 프로그램의 진입점
int main() {
    string youtubeUrl = "https://www.youtube.com/watch?v=2t1i5LYoZXo&list=PL26w7arZdurV-lsFxCVL9zd9j9zrPsM4A";
    string videoUrl = getYouTubeStreamUrl(youtubeUrl);  // YouTube 스트림 URL 가져오기

    if (!videoUrl.empty()) {
        cout << "Stream URL: " << videoUrl << endl;
        processVideo(videoUrl);  // 비디오 스트림 처리
    } else {
        cerr << "비디오 URL을 가져오는 데 실패했습니다." << endl;
    }

    return 0;
}
